kymatio/scattering2d/tests/test_torch_scattering2d.py

```diff
@@ -0,0 +1,380 @@
+class TestCDGMM:
+    @pytest.fixture(params=(False, True))
+    def data(self, request):
+        real_filter = request.param
+        x = torch.rand(100, 128, 128, 2).float()
+        filt = torch.rand(128, 128, 2).float()
+        y = torch.ones(100, 128, 128, 2).float()
+        if real_filter:
+            filt[..., 1] = 0
+        y[..., 0] = x[..., 0] * filt[..., 0] - x[..., 1] * filt[..., 1]
+        y[..., 1] = x[..., 1] * filt[..., 0] + x[..., 0] * filt[..., 1]
+        if real_filter:
+            filt = filt[..., :1]
+        return x, filt, y
+
+    @pytest.mark.parametrize("backend", backends)
+    @pytest.mark.parametrize("device", devices)
+    @pytest.mark.parametrize("inplace", (False, True))
+    def test_cdgmm_forward(self, data, backend, device, inplace):
+        if device == 'cpu' and backend.name == 'torch_skcuda':
+            pytest.skip("skcuda backend can only run on gpu")
+        x, filt, y = data
+        # move to device
+        x, filt, y = x.to(device), filt.to(device), y.to(device)
+        z = backend.cdgmm(x, filt, inplace=inplace)
+        # compare
+        assert torch.allclose(y, z, atol=1e-7, rtol =1e-6) # There is a very small meaningless difference for skcuda+GPU
+
+    @pytest.mark.parametrize("backend", backends)
+    def test_cdgmm_exceptions(self, backend):
+        with pytest.raises(RuntimeError) as exc:
+            backend.cdgmm(torch.empty(3, 4, 5, 2), torch.empty(4, 3, 2))
+        assert "not compatible" in exc.value.args[0]
+        with pytest.raises(TypeError) as exc:
+            backend.cdgmm(torch.empty(3, 4, 5, 1), torch.empty(4, 5, 1))
+        assert "input must be complex" in exc.value.args[0]
+        with pytest.raises(TypeError) as exc:
+            backend.cdgmm(torch.empty(3, 4, 5, 2), torch.empty(4, 5, 3))
+        assert "filter must be complex or real" in exc.value.args[0]
+        with pytest.raises(RuntimeError) as exc:
+            backend.cdgmm(torch.empty(3, 4, 5, 2), torch.empty(3, 4, 5, 2))
+        assert "filter must be a 3-tensor" in exc.value.args[0]
+        with pytest.raises(RuntimeError) as exc:
+            backend.cdgmm(torch.empty(3, 4, 5, 2), torch.empty(4, 5, 1).double())
+        assert "must be of the same dtype" in exc.value.args[0]
+        if 'cuda' in devices:
+            with pytest.raises(RuntimeError) as exc:
+                backend.cdgmm(torch.empty(3, 4, 5, 2), torch.empty(4, 5, 1).cuda())
+            assert "type" in exc.value.args[0]
+
+class TestFFT:
+    @pytest.mark.parametrize("backend", backends)
+    def test_fft(self, backend):
+        x = torch.rand(4, 4, 1)
+        with pytest.raises(TypeError) as record:
+            backend.fft(x)
+        assert ('complex' in record.value.args[0])
+
+        x = torch.randn(4, 4, 2)
+        y = x[::2, ::2]
+
+        with pytest.raises(RuntimeError) as record:
+            backend.fft(y)
+        assert ('must be contiguous' in record.value.args[0])
+
+
+class TestScattering2D_Torch:
```

Comment:
remove underscore

Reply:
how is this resolved? This naming scheme hurts any reasonable person's eyes

Reply:
my bad. I'm writing the tf backend, it's solved on my local branch